### PR TITLE
Fix LSN in keepalive messages, if no WAL has been sent yet

### PR DESCRIPTION
**Pull request by hlinnaka**
When a new connection is established to the safekeeper, the 'end_pos' field is initially set to Lsn::INVALID (i.e 0/0). If there is no WAL to send to the client, we send KeepAlive messages with Lsn::INVALID. That confuses the pageserver: it thinks that safekeeper is lagging very much behind the tip of the branch, and will reconnect to a different safekeeper. Then the same thing happens with the new safekeeper, until some WAL is streamed which sets 'end_pos' to a valid value.

To fix, use 'start_pos' rather than 'end_pos' in the keepalive messages. When the safekeeper has sent all the WAL it has available, they are equal. When the safekeeper has some WAL to send, it will send an XLogData message rather than KeepAlive. If it did send a KeepAlive even when there was some WAL to send too, I think 'start_pos' was a more correct value anyway.

Fixes https://github.com/neondatabase/neon/issues/3972

### CODE REVIEW

File: safekeeper/src/send_wal.rs [hunk 1]
```diff
@@ -551,7 +551,7 @@ impl<IO: AsyncRead + AsyncWrite + Unpin> WalSender<'_, IO> {
 
             self.pgb
                 .write_message(&BeMessage::KeepAlive(WalSndKeepAlive {
-                    sent_ptr: self.end_pos.0,
+                    sent_ptr: self.start_pos.0,
```
Comment (petuhovskiy) [inline safekeeper/src/send_wal.rs:554]: Let's add a comment that `self.start_pos.0` is equal to `commit_lsn` in this case.

**Review by petuhovskiy — APPROVED** [review]
In the pageserver we have this code to process the message:

https://github.com/neondatabase/neon/blob/3ceef7b17aa2452d55a4726aeba94050f9bd551e/pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs#L237-L240

We expect LSN to be `commit_lsn`, aka `The current end of WAL on the server.` (from the postgres docs). `end_pos` indeed can be `Lsn::INVALID`, and `wait_for_lsn` works that way that `start_pos` should be equal to `commit_lsn`, so the fix looks good to me.

I think `WalSndKeepAlive::sent_ptr` is a misleading name, it should be called `wal_end`, the same as in rust-postgres client.

**Comment by hlinnaka** [comment]
I had second thoughts on my earlier fix. Using 'start_pos' in the KeepAlive message isn't quite right, if we ever sent a KeepAlive message when `start_pos != end_pos`. We don't, so we could just add a comment to explain it. But on second thoughts, I think a better fix is to initialize `end_pos` to the starting position, when the streaming starts. I updated this PR to do that instead. Please take another look.

**Comment by hlinnaka** [comment]
> I think WalSndKeepAlive::sent_ptr is a misleading name, it should be called wal_end, the same as in rust-postgres client.

Agreed. Let's fix that as a separate PR.

**Comment by hlinnaka** [comment]
This was superseded by PR #4144, closing.

**Pull request closed without merging**